ion/reader.go

```diff
@@ -109,47 +109,49 @@ type Reader interface {
+	BoolValue() (*bool, error)
 
 	// IntSize returns the size of integer needed to losslessly represent the current value
 	// (if that makes sense). It returns an error if the current value is not an Ion int.
 	IntSize() (IntSize, error)
 
-	// IntValue returns the current value as a 32-bit integer (if that makes sense). It
-	// returns an error if the current value is not an Ion integer or requires more than
-	// 32 bits to represent losslessly.
-	IntValue() (int, error)
+	// IntValue returns the current value as a 32-bit integer (if that makes sense). It returns
+	// nil if the value is null. It returns an error if the current value is not an Ion integer
+	// or requires more than 32 bits to represent losslessly.
+	IntValue() (*int, error)
 
-	// Int64Value returns the current value as a 64-bit integer (if that makes sense). It
-	// returns an error if the current value is not an Ion integer or requires more than
-	// 64 bits to represent losslessly.
-	Int64Value() (int64, error)
+	// Int64Value returns the current value as a 64-bit integer (if that makes sense). It returns
+	// nil if the value is null. It returns an error if the current value is not an Ion integer
+	// or requires more than 64 bits to represent losslessly.
+	Int64Value() (*int64, error)
 
 	// Uint64Value returns the current value as an unsigned 64-bit integer (if that makes
-	// sense). It returns an error if the current value is not an Ion integer, is negative,
-	// or requires more than 64 bits to represent losslessly.
-	Uint64Value() (uint64, error)
+	// sense). It returns nil if the value is null. It returns an error if the current value
+	// is not an Ion integer, is negative, or requires more than 64 bits to represent losslessly.
+	Uint64Value() (*uint64, error)
 
-	// BigIntValue returns the current value as a big.Integer (if that makes sense). It
-	// returns an error if the current value is not an Ion integer.
+	// BigIntValue returns the current value as a big.Integer (if that makes sense). It returns
+	// nil if the value is null. It returns an error if the current value is not an Ion integer.
 	BigIntValue() (*big.Int, error)
 
-	// FloatValue returns the current value as a 64-bit floating point number (if that
-	// makes sense). It returns an error if the current value is not an Ion float.
-	FloatValue() (float64, error)
+	// FloatValue returns the current value as a 64-bit floating point number (if that makes
+	// sense). It returns nil if the value is null. It returns an error if the current value
+	// is not an Ion float.
+	FloatValue() (*float64, error)
 
-	// DecimalValue returns the current value as an arbitrary-precision Decimal (if that
-	// makes sense). It returns an error if the current value is not an Ion decimal.
+	// DecimalValue returns the current value as an arbitrary-precision Decimal (if that makes
+	// sense). It returns nil if the value is null. It returns an error if the current value is
+	// not an Ion decimal.
```

Comment:
Do we have a bound on the Decimal's size that we (our code) error out on? If the error in the case of something too large is from the language that's ok. I am asking if we perform any checks on top of that and generate an error ourselves.

Reply:
Short answer is no.
bigint does not have an explicit limit.

> There is no explicit limit. The limit will be your system resources (memory) allow.

[here](https://ispycode.com/Blog/golang/2016-10/What-is-the-max-value-of-bigint) and [here](https://stackoverflow.com/questions/17564335/golang-math-big-what-is-the-max-value-of-big-int#:~:text=There%20is%20no%20explicit%20limit,where%2010%5E100000%20is%20benchmarked.) for reference.